### ISSUE TITLE
Prevent staff and AH from seeing each others in progress RUPs (Privacy feature).

### ISSUE THREAD

**Issue by LisaMoore1** [bcgov/range-web#683]
EDIT:  Hi Mike here, @LisaMoore1 skip to the bottom to see what to test. : )

@micheal-w-wells can you take a look at this while reviewing the other versioning stuff or forward on to Caleb? Thanks
I’ve put a bug label on this as I think it is a bug but since I am not able to see what is happening with the different versions. Will need someone else to confirm whether this aspect of versions has been addressed or not.

Part of our relationship is that if an AH is working at adding content staff cannot see it until the AH actually chooses to submit it to staff. 
Same applies to amendments — until amendments are signed and submitted staff should not be able to view content
AHs should not be able to view any content staff are working on in amendments or drafts until staff submit it to the AH. 

Basically, if a version is an an “edit” mode for the other user I (the person logged in) should not be able to see the real time content (what has been changed since I last had control). Note that you can see this intent in the headers that are in place indicating to AH that their content will not be visible to staff until they submit.

**Comment by micheal-w-wells** [comment]
@calebissharp relevant to what we just talked about.  I think the thing to do here is:

- snapshot before changing hands (even if not approved, but mark it as such somehow so it doesn't show up in list of versions later)
- when changing hands - make the view only side see the last snapshot

**Comment by calebissharp** [comment]
@micheal-w-wells That doesn't sound too difficult. I wonder though, wouldn't this have always been an issue, before all this versioning stuff?

**Comment by micheal-w-wells** [comment]
@calebissharp it totally would have been, yes.

**Comment by micheal-w-wells** [comment]
@calebissharp I'm trying to figure out where to track this.  I was thinking of adding `is_staff_view` and `is_AH_view` to `plan_snapshot`.  If I'm staff and `is_staff_view` isn't checked for any version then I'm in edit mode.

What do you think?

**Comment by calebissharp** [comment]
@micheal-w-wells Since we already track the status of the plan (`status_id`) with the snapshot we should just be able to use that to deduce which roles should be able to access it. 

We'd have to update the `GET /plan/:id` endpoint to take into account the user role, and that if the current plan is in a status they can't view, it would send the most recent snapshot that they can view. We'd also have to change the agreement search endpoint, since that sends back the plan as well.

**Comment by micheal-w-wells** [comment]
I'll work on `GET /plan/:id`

**Comment by micheal-w-wells** [comment]
I'm struggling a little bit with fetching snapshots

**Comment by micheal-w-wells** [comment]
https://github.com/bcgov/range-api/pull/166 works for letting people on see the live version of plan when it is their turn, doesn't yet work for the other case (when it's not their turn and they need to see last version that was theirs).

**Comment by micheal-w-wells** [comment]
@LisaMoore1 @ZoeSimon  all i have left in my court to do on versioning is finish this one, although we could potentially release without it and I can merge the web portion of this one in when i nail it down.  All the back-end meat and potatoes for versioning is done:

* They are created at the right time
* They're legal status and dates are calculated, and it's easy to tell from the data which is legally standing
* who edited them and what status they change from and to etc

This stuff is all available in a database view and an endpoint (and the new view will work in metabase). 

No change to the front end is done for versioning yet, that is the first enhancement for Oamar.  That one will include filtering out the non-approved / non-legal versions from the list.

**Comment by micheal-w-wells** [comment]
```
Current  plans status                             | Which version to use as 'privacy version' | Who is it for?
----+------+--------------------------------------+                                           +
  1 | C    | Created                              | Most recent 6 (Staff Draft)               | Staff
                                                                                               
  2 | O    | Completed                            |                                           |
  3 | P    | Pending                              |                                           |
  4 | D    | Draft                                |                                           |
  5 | R    | Change Requested                     |                                           |
                                                                                               
  6 | SD   | Staff Draft                          | None, show a message                      | AH
                                                                                               
  7 | WM   | Wrongly Made - Without Effect        | None                                      |
                                                                                               
  8 | SW   | Stands - Wrongly Made                | None                                      |
                                                                                               
  9 | S    | Stands                               | None                                      |
                                                                                               
 10 | NF   | Not Approved - Further Work Required | None                                      |
                                                                                               
 11 | NA   | Not Approved                         | None                                      |
                                                                                               
 12 | A    | Approved                             | None                                      |
                                                                                               
 13 | SR   | Submitted For Review                 | Most recent 1 (Created/sub to ah)         | AH
 14 | SFD  | Submitted For Final Decision         |                                           |
 15 | RR   | Recommend Ready                      |                                           |
 16 | RNR  | Recommend Not Ready                  |                                           |
 17 | RFD  | Ready For Final Decision             |                                           |
 18 | AC   | Awaiting Confirmation                |                                           |
 19 | RFS  | Recommended For Submission           |                                           |
                                                                                               
 20 | MSR  | Stands - Review                      | None                                      |

```


@LisaMoore1 I need help figuring out which version of the plan is appropriate to use as the privacy version and when that is applicable.  I tried to do it in a more generic way based on last person to touch it but i don't think that works in all cases or is applicable in all cases.  If you can help me populate the two columns to the write (can just write as comments and I'll update this) that would be a huge help.

**Comment by LisaMoore1** [comment]
@micheal-w-wells Can you provide me with a little more information on what you mean by “privacy version”?

**Comment by micheal-w-wells** [comment]
Sure!

Scenarios:

- I'm staff.  I make a plan. Save it. It's in staff draft.  I have the 'live working copy of plan'.
- I'm AH.  I shouldn't see what staff is working on.  As I haven't worked on anything or submitted anything yet, there is no 'privacy version' to display while staff work on a draft, so I'll see a message saying it's being. worked on instead of the 'live working copy of the plan'.


- I'm staff.  I submit to AH.  My most recent Staff draft is saved as a version (not one you see in the legal list Oamar is working on, but its in the db).  I should see the content of my draft as the 'privacy version' while AH do their thing, while the header should be smart and show that it is Submitted and not still in Staff Draft.
- I'm AH, i get the live working copy of the plan while I make my changes.  


Does that make sense @LisaMoore1  ?  Sorry if i haven't been clear on this.  We can chat on phone if that is easier.

**Comment by micheal-w-wells** [comment]
![image](https://user-images.githubusercontent.com/955467/80254482-1d848e00-8630-11ea-904c-2a8ac25d62a4.png)

**Comment by micheal-w-wells** [comment]
@calebissharp do you have a branch with those new statuses mentioned in the spreadsheet?

**Comment by micheal-w-wells** [comment]
leaving this here so i can pick up where i left off in the morning :
```
Current  plans status                             | Which version to use as 'privacy version' | Who is it for?
----+------+--------------------------------------+                                           +
INITIATE:
  6 | SD   | Staff Draft                          | None, show a message                      | AH
  1 | C    | Created                              | Most recent 6 (Staff Draft)               | Staff

REVIEW:
 13 | SR   | Submitted For Review                 | Most recent 1 (Created/sub to ah)         | AH
  5 | R    | Change Requested                     | Most recent 13                            | Staff
 19 | RFS  | Recommended For Submission           | Most recent 13                            | Staff

SUBMITTED FOR DECISION:
 18 | AC   | Awaiting Confirmation                |                                           |
 14 | SFD  | Submitted For Final Decision         |                                           |
5 again.
 15 | RR   | Recommend Ready                      |                                           |
 16 | RNR  | Recommend Not Ready                  |                                           |


AMENDMENT INITIATION:
? Amendment initiated (no number yet)             | None, show current approved plan          | either
13 again.
14 again.
1 again.

SUBMITTED FOR DETERMINATION:
  9 | S    | Stands                               | None                                      |
 20 | MSR  | Stands - Review                      | None                                      |


DETERMINATION RECORDED (MINOR AMEND.):
  7 | WM   | Wrongly Made - Without Effect        | None                                      |
  8 | SW   | Stands - Wrongly Made                | None                                      |


DECISION RECORDED (FOR AMENDMENT):
 12 | A    | Approved                             | None                                      |
 11 | NA   | Not Approved                         | None                                      |
 10 | NF   | Not Approved - Further Work Required | None                                      |

7 again.
8 again.
? for Submit  as mandatory


the graveyard:
 17 | RFD  | Ready For Final Decision             | Not used anymore                          |
  2 | O    | Completed                            | Not used anymore                          |
  3 | P    | Pending                              | Not used anymore                          |
  4 | D    | Draft                                | Not used anymore                          |


```

**Comment by micheal-w-wells** [comment]
Here's the gist of how it works:

- As the RUP changes statuses, we take snapshots/versions at each stage
- When you aren't supposed to see it, the database has a version marked as 'StaffView' or 'AHView' when that is the case, based on the spreadsheet you gave me.
- The site will load the old content of the plan, but will use the current statuses so the headers and buttons still make sense.

I still have to make it work for the new amendment statuses which will be a new small issue,  but the bulk of it is done and you can test it out and see it work and find kinks if any while I finish translating your spreadsheet to my query on Monday.

To test:
- Be staff.  Make a draft.  Set some date on the schedule and make a note of it.
- Submit it.  See the headers still work.
- Be AH. Change the date.  Save.
- Be staff. You should still see your original. dates. etc on the plan.


This logic carries forward through the workflow up to approved RUP.  I'll link the amendment update soon as its done.

**Comment by LisaMoore1** [comment]
@micheal-w-wells. I’ve passed an amendment back and forth a few times with something text changes and this seems to be working. I’m closing.
